allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+
+T = TypeVar('T')
+
+# getattr is necessary because mypy insists that no such attribute exists.
+_NO_DEFAULT = getattr(inspect, '_empty')  # pylint: disable=invalid-name
+
+def takes_arg(obj, arg: str) -> bool:
+    """
+    Checks whether the provided obj takes a certain arg.
+    If it's a class, we're really checking whether its constructor does.
+    If it's a function or method, we're checking the object itself.
+    Otherwise, we raise an error.
+    """
+    if inspect.isclass(obj):
+        signature = inspect.signature(obj.__init__)
+    elif inspect.ismethod(obj) or inspect.isfunction(obj):
+        signature = inspect.signature(obj)
+    else:
+        raise ConfigurationError(f"object {obj} is not callable")
+    return arg in signature.parameters
+
+def remove_optional(annotation):
+    """
+    Optional[X] annotations are actually represented as Union[X, NoneType].
+    For our purposes, the "Optional" part is not interesting, so here we
+    throw it away.
+    """
+    origin = getattr(annotation, '__origin__', None)
+    args = getattr(annotation, '__args__', ())
+    if origin == Union and len(args) == 2 and args[1] == type(None):
+        return args[0]
+    else:
+        return annotation
+
+def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
+    """
+    Given some class, a `Params` object, and potentially other keyword arguments,
+    create a dict of keyword args suitable for passing to the class's constructor.
+    """
+    # Get the signature of the constructor.
+    signature = inspect.signature(cls.__init__)
+    kwargs: Dict[str, Any] = {}
+
+    # Iterate over all the constructor parameters and their annotations.
+    for name, param in signature.parameters.items():
+        # Skip "self". You're not *required* to call the first parameter "self",
+        # so in theory this logic is fragile, but if you don't call the self parameter
+        # "self" you kind of deserve what happens.
+        if name == "self":
+            continue
+
+        # Get the annotation and break it into `origin` and `args`
+        annotation = remove_optional(param.annotation)
+        origin = getattr(annotation, '__origin__', None)
+        args = getattr(annotation, '__args__', [])
+        default = param.default
+        optional = default != _NO_DEFAULT
+
+        # Some constructors expect extra non-parameter items, e.g. vocab: Vocabulary.
+        # We check the provided `extras` for these and just use them if they exist.
+        if name in extras:
+            kwargs[name] = extras[name]
+
+        # The next case is when the parameter type is itself constructible from_params.
+        elif hasattr(annotation, 'from_params'):
+            if name in params:
+                # Our params have an entry for this, so we use that.
+                subparams = params.pop(name)
+
+                if takes_arg(annotation.from_params, 'extras'):
```

Comment:
I don't understand this one.  When do we want to pass on extras?  I'm pretty sure this is the mechanism you're using for `Vocab` and such, but I'm not sure when you want to pass that on...  Oh, do we sometimes pass the `Vocab` down?  Is that what this is about?

Reply:
yes, for example, `BasicTextFieldEmbedder.from_params` is custom and needs a Vocab, but the base `TextFieldEmbedder.from_params` doesn't

Reply:
I feel like that was a bug with `TextFieldEmbedder` - the API would break if the `from_params` method differed between the base class and subclasses.  Yes?

Reply:
you could call it a bug, or you could call it a design error, but yes the text_field_embedder section of a config probably should have looked like

```
  "model": {
    "type": "bidaf",
    "text_field_embedder": {
      "token_embedders": {
          "tokens": { ... },
          "token_characters": { ... },
         }
      }
    },
```

instead of

```
  "model": {
    "type": "bidaf",
    "text_field_embedder": {
      "tokens": { ... },
      "token_characters": { ... },
      }
    },
```

but we can't really change it now without breaking all of the model configs

Reply:
Oh, I was talking about something else, I think.  You said the `BasicTextFieldEmbedder.from_params` needs `vocab`, but `TextFieldEmbedder.from_params` doesn't.  Actually, they both do, currently, so we didn't mess up the API.  I think I understand why passing along `extras` is necessary now, so my original comment here isn't a problem.

What you mention in your most recent comment is a bit of a pain, though, I agree.  Could we migrate to the cleaner version with a deprecation warning?  That is, accept both of them, and if we get a config file with the current format, emit a warning?  (Also not necessary for this PR.)